File: mobile/test/commit_tracker_test.dart
import "package:beresta/commit_tracker.dart";
import "package:flutter_test/flutter_test.dart";

void main() {
  group("CommitTracker", () {
    test("issues strictly increasing generations", () {
      final tracker = CommitTracker();
      var previous = 0;
      for (var i = 0; i < 5; i++) {
        final current = tracker.dirty();
        expect(current, greaterThan(previous));
        previous = current;
      }
    });

    test("current() does not itself mark the session dirty", () {
      final tracker = CommitTracker();
      final generation = tracker.dirty();
      expect(tracker.current(), generation);
      expect(tracker.current(), generation);
    });

    test("accepts a success for the latest generation as saved", () {
      final tracker = CommitTracker();
      final generation = tracker.dirty();
      expect(tracker.accept(generation, true), LocalSaveState.saved);
    });

    test("accepts a failure for the latest generation as couldNotSave", () {
      final tracker = CommitTracker();
      final generation = tracker.dirty();
      expect(tracker.accept(generation, false), LocalSaveState.couldNotSave);
    });

    // The scenario design.md decision 2 requires: a slow commit for an
    // older generation must not report saved once newer, still-unsaved
    // input has arrived, even though that older commit itself succeeded.
    test(
      "discards a stale success instead of marking newer dirty content as saved",
      () {
        final tracker = CommitTracker();
        final stale = tracker.dirty();
        tracker.dirty(); // newer input arrives before the stale commit resolves

        expect(tracker.accept(stale, true), isNull);
      },
    );

    test(
      "discards a stale failure instead of flagging newer dirty content as failed",
      () {
        final tracker = CommitTracker();
        final stale = tracker.dirty();
        tracker.dirty();

        expect(tracker.accept(stale, false), isNull);
      },
    );

    test(
      "still reports the latest generation correctly after a stale completion was discarded",
      () {
        final tracker = CommitTracker();
        final stale = tracker.dirty();
        final latest = tracker.dirty();

        expect(tracker.accept(stale, true), isNull);
        expect(tracker.accept(latest, true), LocalSaveState.saved);
      },
    );

    test("re-affirms saved on a duplicate success for the same generation", () {
      final tracker = CommitTracker();
      final generation = tracker.dirty();

      expect(tracker.accept(generation, true), LocalSaveState.saved);
      expect(tracker.accept(generation, true), LocalSaveState.saved);
    });

    test(
      "does not let a failure regress a generation an earlier attempt already saved",
      () {
        final tracker = CommitTracker();
        final generation = tracker.dirty();

        expect(tracker.accept(generation, true), LocalSaveState.saved);
        expect(tracker.accept(generation, false), isNull);
      },
    );

    test(
      "accepts a retry's success for the same generation after an earlier failure",
      () {
        // The common retry path: a commit fails, and since nothing new was
        // typed in between, the retry captures the exact same generation
        // as the failed attempt.
        final tracker = CommitTracker();
        final generation = tracker.dirty();

        expect(tracker.accept(generation, false), LocalSaveState.couldNotSave);
        expect(tracker.accept(generation, true), LocalSaveState.saved);
      },
    );

    test(
      "discards a late out-of-order completion of an earlier in-flight commit",
      () {
        // Two commits can be in flight if a debounce commits generation 1,
        // more input arrives (generation 2), and generation 2's commit
        // completes before generation 1's slower one settles.
        final tracker = CommitTracker();
        final first = tracker.dirty();
        final second = tracker.dirty();

        expect(tracker.accept(second, true), LocalSaveState.saved);
        expect(tracker.accept(first, true), isNull);
      },
    );

    // A real bug found while wiring this into the editor commit
    // controller: a commit with no prior dirty() call targets generation
    // 0, a legitimate first attempt, not a stale or already-resolved
    // duplicate.
    test(
      "does not discard a failure at generation 0 with no prior dirty() call",
      () {
        final tracker = CommitTracker();
        expect(tracker.accept(0, false), LocalSaveState.couldNotSave);
      },
    );

    test(
      "does not discard a success at generation 0 with no prior dirty() call",
      () {
        final tracker = CommitTracker();
        expect(tracker.accept(0, true), LocalSaveState.saved);
      },
    );
  });
}

File: desktop/frontend/src/editor/commitTracker.test.ts
import { describe, expect, it } from "vitest";

import { CommitTracker } from "./commitTracker";

describe("CommitTracker", () => {
  it("issues strictly increasing generations", () => {
    const tracker = new CommitTracker();
    let previous = 0;
    for (let i = 0; i < 5; i++) {
      const current = tracker.dirty();
      expect(current).toBeGreaterThan(previous);
      previous = current;
    }
  });

  // A real bug found while wiring this into the editor commit controller:
  // a commit with no prior dirty() call (a title-only rename before any
  // body edit) targets generation 0, a legitimate first attempt, not a
  // stale or already-resolved duplicate.
  it("does not discard a failure at generation 0 with no prior dirty() call", () => {
    const tracker = new CommitTracker();
    expect(tracker.accept(0, false)).toBe("could_not_save");
  });

  it("does not discard a success at generation 0 with no prior dirty() call", () => {
    const tracker = new CommitTracker();
    expect(tracker.accept(0, true)).toBe("saved");
  });

  it("current() does not itself mark the session dirty", () => {
    const tracker = new CommitTracker();
    const generation = tracker.dirty();
    expect(tracker.current()).toBe(generation);
    expect(tracker.current()).toBe(generation);
  });

  it("accepts a success for the latest generation as saved", () => {
    const tracker = new CommitTracker();
    const generation = tracker.dirty();
    expect(tracker.accept(generation, true)).toBe("saved");
  });

  it("accepts a failure for the latest generation as could_not_save", () => {
    const tracker = new CommitTracker();
    const generation = tracker.dirty();
    expect(tracker.accept(generation, false)).toBe("could_not_save");
  });

  // The scenario design.md decision 2 requires: a slow commit for an
  // older generation must not report "saved" once newer, still-unsaved
  // input has arrived, even though that older commit itself succeeded.
  it("discards a stale success instead of marking newer dirty content as saved", () => {
    const tracker = new CommitTracker();
    const stale = tracker.dirty();
    tracker.dirty(); // newer input arrives before the stale commit resolves

    expect(tracker.accept(stale, true)).toBeNull();
  });

  it("discards a stale failure instead of flagging newer dirty content as failed", () => {
    const tracker = new CommitTracker();
    const stale = tracker.dirty();
    tracker.dirty();

    expect(tracker.accept(stale, false)).toBeNull();
  });

  it("still reports the latest generation correctly after a stale completion was discarded", () => {
    const tracker = new CommitTracker();
    const stale = tracker.dirty();
    const latest = tracker.dirty();

    expect(tracker.accept(stale, true)).toBeNull();
    expect(tracker.accept(latest, true)).toBe("saved");
  });

  it("re-affirms saved on a duplicate success for the same generation", () => {
    const tracker = new CommitTracker();
    const generation = tracker.dirty();

    expect(tracker.accept(generation, true)).toBe("saved");
    expect(tracker.accept(generation, true)).toBe("saved");
  });

  it("does not let a failure regress a generation an earlier attempt already saved", () => {
    const tracker = new CommitTracker();
    const generation = tracker.dirty();

    expect(tracker.accept(generation, true)).toBe("saved");
    expect(tracker.accept(generation, false)).toBeNull();
  });

  it("accepts a retry's success for the same generation after an earlier failure", () => {
    // The common retry path: a commit fails, its payload is requeued for
    // the next attempt, and since nothing new was typed in between, the
    // retry captures the exact same generation as the failed attempt.
    const tracker = new CommitTracker();
    const generation = tracker.dirty();

    expect(tracker.accept(generation, false)).toBe("could_not_save");
    expect(tracker.accept(generation, true)).toBe("saved");
  });

  it("discards a late out-of-order completion of an earlier in-flight commit", () => {
    // Two commits can be in flight if a debounce commits generation 1,
    // more input arrives (generation 2), and generation 2's commit
    // completes before generation 1's slower one settles.
    const tracker = new CommitTracker();
    const first = tracker.dirty();
    const second = tracker.dirty();

    expect(tracker.accept(second, true)).toBe("saved");
    expect(tracker.accept(first, true)).toBeNull();
  });
});
